ui-admin/src/portal/AddMailingListUsersModal.tsx

```diff
@@ -0,0 +1,161 @@
+    onClose: () => void,
+    reload: () => void
+}) {
+  const [contacts, setContacts] = useState<MailingListContact[]>([{ name: '', email: '' }])
+  const [isLoading, setIsLoading] = useState(false)
+  const { FileChooser } = useFileUploadButton(file => {
+    const reader = new FileReader()
+    reader.onload = () => {
+      setContacts(parseContactCsv(reader.result as string))
+    }
+    reader.readAsText(file)
+  }, 'Import CSV')
+
+  //A contact is invalid if it has a name but no email, or an email that doesn't match the email regex.
+  //Contacts without names are allowed.
+  const hasInvalidContacts = contacts.some(contact =>
+    (!contact.email && contact.name) || (contact.email && !contact.email.match(/.+@.+\..+/)))
+
+  const addUsers = async () => {
+    setIsLoading(true)
+    const nonEmptyContacts = contacts.filter(contact => contact.email && contact.name)
+    try {
+      const response = await Api.addMailingListContacts(
+        portalContext.portal.shortcode, portalEnv.environmentName, nonEmptyContacts
+      )
+      Store.addNotification(successNotification(`${response.length} new ${pluralize('user', response.length)} added`))
+      onClose()
+    } catch {
+      Store.addNotification(failureNotification('Error: could not add users'))
+    }
+    reload()
+    setIsLoading(false)
+  }
+
+  return <Modal show={show} className="modal-lg" onHide={onClose}>
+    <Modal.Header closeButton>
+      <Modal.Title>Add Users</Modal.Title>
+    </Modal.Header>
+    <Modal.Body>
+      <div className="pb-3">
+                Add users to your mailing list by entering their name and email address. You
+                can also import a <code>.csv</code> file with multiple contacts. The file should
+                be formatted as <code>name,emailAddress</code>.
+      </div>
+      <table className="ms-2 table">
+        <thead>
+          <tr>
+            <td className="fw-semibold">Name</td>
+            <td className="fw-semibold">Email Address*</td>
+            <td></td>
+          </tr>
+        </thead>
+        <tbody>
+          {contacts.map((contact, i) =>
+            <ContactEntry
+              key={i}
+              contact={contact}
+              onChange={newContact => {
+                setContacts(emails => [
+                  ...emails.slice(0, i),
+                  { ...emails[i], ...newContact },
+                  ...emails.slice(i + 1)
+                ])
+              }}
+              onRemove={() =>
+                setContacts(emails => [
+                  ...emails.slice(0, i),
+                  ...emails.slice(i + 1)
+                ])
+              }
+            />
+          )}
+        </tbody>
+      </table>
+      <button className="btn btn-primary"
+        onClick={() =>
+          setContacts([...contacts, { name: '', email: '' }])
+        }>
+        Add
+      </button>
+    </Modal.Body>
+    <Modal.Footer>
+      <LoadingSpinner isLoading={isLoading}>
+        {FileChooser}
```

Comment:
I think it's confusing to have this button here.  How about directly underneath the "Add" button?

Reply:
Done!

<img width="812" alt="Screenshot 2024-04-01 at 2 13 30 PM" src="https://github.com/broadinstitute/juniper/assets/7257391/18f8befa-9aca-4ea5-949a-1fdd1ee7b5ae">